src/core/unif_rule.ml

```diff
@@ -71,3 +71,37 @@ let rec p_unpack : p_term -> (p_term * p_term) list = fun eqs ->
       else if id s = "#equiv" then [(v, w)] else
       assert false (* Ill-formed term. *)
   | _                               -> assert false (* Ill-formed term. *)
+
+let prod : sym =
+  Sign.add_symbol sign Public Defin (Pos.none "#prod") Kind []
+
+let rec prod_obj : term -> term = fun t ->
+  match unfold t with
+  | Appl(t,u) -> Appl(prod_obj t, prod_obj u)
+  | Prod(a,b) ->
+      let (x, b) = Bindlib.unbind b in
+      let b = lift (prod_obj b) in
+      let b = Bindlib.unbox (Bindlib.bind_var x b) in
+      Appl(Symb(prod), Abst(prod_obj a,b))
+  | Abst(a,b)         ->
+      let (x, b) = Bindlib.unbind b in
+      let b = lift (prod_obj b) in
+      let b = Bindlib.(unbox (bind_var x b)) in
+      Abst(prod_obj a, b)
+  | _                 -> t
+
+let rec prod_unobj : term -> term = fun t ->
+  match unfold t with
+  | Appl(Symb(s), Abst(a,b)) when s == prod ->
+      let (x, b) = Bindlib.unbind b in
+      let b = lift (prod_unobj b) in
+      let b = Bindlib.unbox (Bindlib.bind_var x b) in
+      Prod(prod_unobj a, b)
+  | Appl(t, u)                              ->
+      Appl(prod_unobj t, prod_unobj u)
+  | Abst(a, b)                              ->
+      let (x, b) = Bindlib.unbind b in
+      let b = lift (prod_unobj b) in
+      let b = Bindlib.unbox (Bindlib.bind_var x b) in
+      Abst(prod_unobj a, b)
+  | _                                       -> t
```

Comment:
The LLet case is missing (may occur in a RHS).

Reply:
Since we rewrite step by step, do we really need prod_unobj? Only the top product(s?) of the LHS and of the term to rewrite needs to be objectified.